### ISSUE TITLE
Control panel security improvements

### ISSUE THREAD

**Issue by gompster** [mail-in-a-box/mailinabox#697]
Hi,

I would like to suggest some security improvements to the control panel web application:

* Add the ability to restrict access to the control panel based on a whitelist of IP addresses (which could be configured in the control panel itself). An administrative interface should not be publicly accessible.
* Use a generic 'login failed' message instead of giving an attacker information about whether a username exists in the system or not.
* Implement protection against brute-force password guessing attacks, preferably by requiring the user to solve a CAPTCHA after ~5 failed login attempts.
* Separate the login page from the rest of the application so that the URL structure (or available functionality) of the application is not revealed to an unauthenticated user.

**Comment by JoshData** [comment]
> Implement protection against brute-force password guessing attacks, preferably by requiring the user to solve a CAPTCHA after ~5 failed login attempts.

It would be nice if this were managed by fail2ban so we have a consistent policy across services.

> Finally you may want to implement additional protection mechanisms against cross-site request forgery attacks.

Do you see a vulnerability or is this hypothetical?

**Comment by aspdye** [comment]
We could just create a "jail" for Fail2Ban so that Roundcube, ownCloud AND the Admin Panel would be secured :wink:

**Comment by gompster** [comment]
Fail2ban integration sound like a good plan, but beware of undesired lockouts.

I dived a little deeper in the session handling in the control panel. It seems cross-site request forgery is not possible because the target user's browser will not append the Authorization HTTP header automatically to the forged request.

There is however one other issue that comes to mind when storing credentials (or some form of credentials) in the local storage. Please beware that every other page/application on the same domain as the control panel will be able to access local storage information. Should other applications contain vulnerabilities that allow an attacker to execute JavaScript (cross-site scripting vulnerabilities), the attacker is able to obtain the control panel's credentials and use them in his own browser. Using cookies instead of the local storage would be an improvement, since cookies can be equipped with a HttpOnly flag that prevents client-side scripts from accessing the cookie value.

**Comment by gompster** [comment]
And while we are at it, another good idea would be to implement protection against clickjacking by adding an `X-Frame-Options` HTTP header to the nginx configuration.

**Comment by aspdye** [comment]
Interesting Aspect! So ownCloud, Roundcube or Static Security Vulnerabilites could "Break" the whole Control Panel and the Mailserver. 

@JoshData Security Priority Bug?!

**Comment by JoshData** [comment]
Fair enough. I was trying to get around CSRF attacks by using local storage, but I see how that opens up another avenue. (Note: CSRF attacks are an _actual_ vulnerability, whereas local storage is a hypothetical problem that relies on a possibly nonexistent vulnerability elsewhere.)

**Comment by gompster** [comment]
So far, I did the following to harden my mail-in-a-box server:

* Harden Postfix SSL configuration by disabling the weak SSLv3 protocol and weak RC4 algorithms:
```
postconf smtpd_tls_mandatory_protocols=\!SSLv2,\!SSLv3
postconf smtpd_tls_protocols=\!SSLv2,\!SSLv3
postconf smtpd_tls_ciphers=high
service postfix restart
```
This results in a better rating at https://ssl-tools.net/mailservers.

* Add protection against clickjacking attacks by adding an `X-Frame-Options` HTTP header. In `/etc/nginx/conf.d/local.conf`, add the following directive to the **server** section:
`add_header X-Frame-Options "SAMEORIGIN";`

* Only allow whitelisted IP addresses to access the control panel. In `/etc/nginx/conf.d/local.conf`, add the following to the **location /admin/** section:
```
allow $MY_IP;
deny all;
```
Replace $MY_IP with the IP address that should have access to the control panel. Other IP addresses that will try to access the control panel will be presented with a HTTP 403: Forbidden response.

It would be nice if users could ultimately configure the IP whitelist from the control panel.
The other two improvements are quick wins. The other improvements in my first post should also be easy to implement. Rebuilding the application to use sessions instead of the local storage would take some more work, so perhaps this is something for the long run.

**Comment by JoshData** [comment]
> Harden Postfix SSL configuration by disabling the weak SSLv3 protocol and weak RC4 algorithms:

See #611 for that.

> Add protection against clickjacking attacks by adding an X-Frame-Options

Submit a PR to add that to https://github.com/mail-in-a-box/mailinabox/blob/master/conf/nginx-primaryonly.conf?

> Rebuilding the application to use sessions instead of the local storage would take some more work

Maybe it's better, at least in the short term, to simply not have persistent logins to the control panel. That would drop localStorage and wouldn't break the architecture at all. The control panel front-end would store the privileges in a javascript variable, rather than in locaStorage.

**Comment by scottnzuk** [comment]
[Login Cross Site Request Forgery CSRF XSRF.pdf](https://github.com/mail-in-a-box/mailinabox/files/1606351/Login.Cross.Site.Request.Forgery.CSRF.XSRF.pdf)

So I run a secury scan on my mail in a box install and this critical security flaw was reported. (hopefully this is the correct place to report)

**Comment by scottnzuk** [comment]
FULL REPORT - All Findings 

[Detectify_FullReport_kiwiz.co.uk_2018-01-04T16.12.01+00_00.pdf](https://github.com/mail-in-a-box/mailinabox/files/1606382/Detectify_FullReport_kiwiz.co.uk_2018-01-04T16.12.01.00_00.pdf)

**Comment by yodax** [comment]
Ideally open a new issue for this. But first look into [this](https://github.com/mail-in-a-box/mailinabox/blob/48ff664ee9393ff0fe85811c12ae0cdd1ced86b7/management/templates/index.html#L339) line of code.

**Comment by remyabel** [comment]
Why not use a dedicated Flask plugin, something like [FlaskSecurity](https://pythonhosted.org/Flask-Security/)? This way we don't reinvent the wheel.

**Comment by timveld** [comment]
There are other issues discussing this topic e.g. #963 
I believe there should be a supported way to tighten security on the web interfaces. At the very least, I should be able to: 
- selectively disable services (e.g. roundcube, which I don't need) 
- Selectively restrict access to services by IP address (e.g. only allow admin access from my VPN) 
Only realistic way to do this is by restricting access to port 443 in ufw, but then I break autodiscover and calendar functionality for my mobile clients. Forcing them all into my VPN is possible but very inconvenient and shouldn't be necessary. 

As a compromise between MIAB's 'no configuration needed' philosophy and the need for customizable security in today's world, there could be some sort of security / advanced options config file which enables me to set these options at install time. This could include other options like disabling pop3 (I currently block the ports in UFW, but the only safe service is a dead service). 

    The argument that 2FA, fail2ban and other measures mean we can safely expose everything to the public internet and keep unused services running no longer holds today. Attacks get more sophisticated and minimizing the attack surface is our best defense. No matter what measures we take, nothing is totally secure. 

Another reason why I strongly feel MIAB should strike a compromise in favor of security is that the world has changed. One reason that people would use MIAB today is that they need to protect themselves against the increasing number of authoritarian regimes. Those regimes are not script kiddies: they know the exploits and will use them to eavesdrop or disrupt services. A determined intelligence agency which explicitly targets someone is hard to stop. But by minimizing the potential for unnecessary backdoors, we can avoid getting on their radar and make it harder for them to quietly read along.

**Comment by JoshData** [comment]
> the world has changed. One reason that people would use MIAB today is that they need to protect themselves against the increasing number of authoritarian regimes. 

This is not a new issue. :) I started this project around the time of the Snowden leaks about government surveillance.

However, I do always like to note that self-hosting is probably a poor choice for someone concerned about government threats. The security teams and the lawyers at Big Email (Gmail, etc.) are going to be far more capable of stopping both attacks and government subpoenas than most of us are.

**Comment by timveld** [comment]
Ah, so you do trust big tech not to willingly leak data?- you're probably right. Also I think I get your point: MiaB is about 'stable independence' rather than obscurity. If a user's primary goal were to stay out of sight, he's better off joining the billions of big tech users amongst whom he won't stand out. Is that what you're saying? 

Ten years ago I would have unambiguously agreed. Now, I do feel that we should get a bit more control over which services are available to which peers on which ports. I mean: who uses POP3 today, So why force everyone to leave it turned on? I get that it is a fine line - customization introduces potential for error - but I'm left with the feeling that MiaB should strike a better balance than it does today.

**Comment by JoshData** [comment]
> Ah, so you do trust big tech not to willingly leak data?

I think you are missing the point or are just wholly naive about the relative power of state actors versus you.
